gordon_janitor_gcp/gce_clients.py

```diff
@@ -0,0 +1,234 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2018 Spotify AB
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Client classes to retrieve project and instance data from GCE.
+
+These clients use the asynchronous HTTP client defined in
+:py:mod:`gordon_janitor_gcp.http_client` and require service account or
+JWT-token credentials for authentication.
+
+To use:
+
+.. code-block:: python
+
+    import asyncio
+
+    import aiohttp
+
+    from gordon_janitor_gcp import auth
+
+    loop = asyncio.get_event_loop()
+
+    async def main():
+        session = aiohttp.ClientSession()
+        loop = asyncio.get_event_loop()
+        auth_client = auth.GoogleAuthClient(k
+            eyfile='/path/to/keyfile', session=session)
+        client = GCPClient(auth_client, session)
+        instances = await client.list_instances('project-id')
+        print(instances)
+
+    loop.run_until_complete(main())
+    # example output
+    # [{'hostname': 'instance-1', 'internal_ip': '10.10.10.10',
+    #   'external_ip': '192.168.1.10'}]
+"""
+
+import logging
+
+from gordon_janitor_gcp import http_client
+
+
+class CRMClient(http_client.AIOGoogleHTTPClient,
+                http_client.PagingGoogleClientMixin):
+    """Async client to interact with Google Cloud Resource Manager API.
+
+    API endpoint documentation:
+    https://cloud.google.com/resource-manager/reference/rest/#rest-resource-v1projects
+
+    Attributes:
+        BASE_URL (str): Base endpoint URL.
+
+    Args:
+        auth_client (gordon_janitor_gcp.auth.GoogleAuthClient): client
+            to manage authentication for HTTP API requests.
+        session (aiohttp.ClientSession): (optional) ``aiohttp`` HTTP
+            session to use for sending requests. Defaults to the
+            session object attached to ``auth_client`` if not provided.
+    """
+    BASE_URL = 'https://cloudresourcemanager.googleapis.com'
+
+    def __init__(self, auth_client=None, session=None):
+        super().__init__(auth_client=auth_client, session=session)
+
+    def parse_response_into_items(self, response, items):
+        """Parse API response for all active projects into a list.
+
+        Args:
+            response (dict): CRM API endpoint response data.
+            items (list): list for collecting parsed results.
+        """
+        projects = response.get('projects', [])
+        for project in projects:
+            if project.get('lifecycleState', '').lower() == 'active':
```

Comment:
I don't see anything weird about it, it looks like a regular ol' mixin class. But since it can cause confusion, I rewrote the pagination mixin and these clients to make 'em less confusing.